artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/QueueControlImpl.java

```diff
@@ -1361,6 +1361,32 @@ public boolean moveMessage(final long messageID,
 
    }
 
+   @Override
+   public boolean copyMessage(final long messageID,
+                              final String targetQueue) throws Exception {
+      // this is a critical task, we need to prevent parallel tasks running
+      try (AutoCloseable lock = server.managementLock()) {
+         if (AuditLogger.isBaseLoggingEnabled()) {
+            AuditLogger.copyMessage(queue, messageID, targetQueue);
+         }
+         checkStarted();
+
+         clearIO();
+         try {
+            Binding binding = server.getPostOffice().getBinding(SimpleString.of(targetQueue));
+
+            if (binding == null) {
+               throw ActiveMQMessageBundle.BUNDLE.noQueueFound(targetQueue);
+            }
+
+            return queue.copyReference(messageID, binding.getAddress(), binding);
```

Comment:
The implementation details here should be part of the Queue::copyReference.


Also... what if the targetQueue was passed a topic.

Shouldn't this call PostOffice::route instead?...

on that case targetQueue here should be targetAddress...




I started writing a test to validate this test and I bumped into that

Reply:
As with the move method this is adding directly to a queue, rerouting it would send it to all queus on an address which is wrong. I could do use route and add rout-to-id of the queue but would there be any benefit to doing that?

Reply:
actually ignore, if you follow the code thru it does use rout on the binding itself so it only goes to that queue